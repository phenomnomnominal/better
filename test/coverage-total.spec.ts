import type { BettererCoverageIssues } from '@betterer/coverage';

import { createFixture } from './fixture';

describe('betterer', () => {
  it('should report the total coverage', async () => {
    const { betterer } = await import('@betterer/betterer');

    const fixture = await createFixture('coverage-total', {
      '.betterer.js': `
const { coverageTotal } = require('@betterer/coverage');

module.exports = {
  test: () => coverageTotal()
};    
        `,
      'tsconfig.json': `
{
  "extends": "../../tsconfig.json",
  "include": ["./src/**/*", "./.betterer.js"]
}
      `
    });

    const coverage = `{  
  "total": {
    "lines": { "total": 220, "covered": 110, "skipped": 0, "pct": 50 },
    "statements": { "total": 410, "covered": 140, "skipped": 0, "pct": 34.1 },
    "functions": { "total": 30, "covered": 11, "skipped": 0, "pct": 36.6 },
    "branches": { "total": 300, "covered": 100, "skipped": 0, "pct": 33.3 }
  }
}`;

    const configPaths = [fixture.paths.config];
    const resultsPath = fixture.paths.results;
    const coveragePath = fixture.resolve('./coverage/coverage-summary.json');

    await fixture.writeFile(coveragePath, coverage);

    const newTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(fixture.testNames(newTestRun.new)).toEqual(['test']);

    const result = await fixture.readFile(resultsPath);

    expect(result).toMatchSnapshot();

    const sameTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(fixture.testNames(sameTestRun.same)).toEqual(['test']);

    const [sameSummary] = sameTestRun.same;
    const testResult = sameSummary.result?.value as BettererCoverageIssues;

    expect(testResult.total).toEqual({
      lines: 220 - 110,
      statements: 410 - 140,
      functions: 30 - 11,
      branches: 300 - 100
    });

    const betterCoverage = `{  
      "total": {
        "lines": { "total": 220, "covered": 210, "skipped": 0, "pct": 95.4 },
        "statements": { "total": 410, "covered": 140, "skipped": 0, "pct": 34.1 },
        "functions": { "total": 30, "covered": 11, "skipped": 0, "pct": 36.6 },
        "branches": { "total": 300, "covered": 100, "skipped": 0, "pct": 33.3 }
      }
    }`;

    await fixture.writeFile(coveragePath, betterCoverage);

    const betterTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(fixture.testNames(betterTestRun.better)).toEqual(['test']);

    const worseCoverage = `{  
      "total": {
        "lines": { "total": 220, "covered": 210, "skipped": 0, "pct": 95.4 },
        "statements": { "total": 410, "covered": 140, "skipped": 0, "pct": 34.1 },
        "functions": { "total": 30, "covered": 11, "skipped": 0, "pct": 36.6 },
        "branches": { "total": 300, "covered":90, "skipped": 0, "pct": 30 }
      }
    }`;

    await fixture.writeFile(coveragePath, worseCoverage);

    const worseTestRun = await betterer({ configPaths, resultsPath, workers: false });

    expect(fixture.testNames(worseTestRun.worse)).toEqual(['test']);

    expect(fixture.logs).toMatchSnapshot();

    await fixture.cleanup();
  });
});
