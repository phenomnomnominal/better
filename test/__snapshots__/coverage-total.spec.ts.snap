// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`betterer should report the total coverage 1`] = `
"// BETTERER RESULTS V2.
//
// If this file contains merge conflicts, use \`betterer merge\` to automatically resolve them:
// https://phenomnomnominal.github.io/betterer/docs/results-file/#merge
//
exports[\`test\`] = {
  value: \`{
  \\"total\\": { \\"lines\\": 110, \\"statements\\": 270, \\"functions\\": 19, \\"branches\\": 200 }
}
\`
};
"
`;

exports[`betterer should report the total coverage 2`] = `
Array [
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ test: \\"test\\" got checked for the first time! 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" got checked for the first time! 🎉
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" got checked for the first time! 🎉

1 test got checked. 🤔
1 test got checked for the first time! 🎉
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ test: \\"test\\" stayed the same. 😐
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" stayed the same. 😐
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" stayed the same. 😐

1 test got checked. 🤔
1 test stayed the same. 😐
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
✅ test: \\"test\\" got better! 😍
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" got better! 😍
",
  "🎉 Betterer (0ms): 1 test done!
✅ test: \\"test\\" got better! 😍

1 test got checked. 🤔
1 test got better! 😍
",
  "🌟 Betterer (0ms):
",
  "🌟 Betterer (0ms): 1 test running...
🤔 test: running \\"test\\"!
",
  "🌟 Betterer (0ms): 1 test running...
🔥 test: \\"test\\" got worse. 😔
",
  "🌟 Betterer (0ms): 1 test running...
🔥 test: \\"test\\" got worse. 😔

Error: \\"test\\" got worse. 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 test: \\"test\\" got worse. 😔
・ \\"branches\\" coverage is worse in \\"total\\": 210 > 200

Error: \\"test\\" got worse. 😔
",
  "💥 Betterer (0ms): 1 test done! 1 test errored!
🔥 test: \\"test\\" got worse. 😔
・ \\"branches\\" coverage is worse in \\"total\\": 210 > 200

Error: \\"test\\" got worse. 😔

1 test got checked. 🤔
1 test got worse. 😔

You should try to fix the new issues! As a last resort, you can run \`betterer --update\` to force an update of the results file. 🆙
",
]
`;
